how-it-works/data-and-provider-model/data-flow-and-lifecycle.md

```diff
@@ -1,37 +1,163 @@
+    D[Set policy for desired<br>BSP replicas]
+    A-->B
+    B-->C
+    C-->D
+```
 
-6. **Ongoing health checks**: On a schedule, the chain challenges BSPs (not MSPs) to prove they still hold your data. BSPs provide small cryptographic proofs; slashing is possible in the event of data loss by BSPs per the protocol rules.
+## Upload the File
 
-7. **Update or replace a file**: Uploading changed content produces a new fingerprint (a new version). The bucket updates its on-chain summary to reference the new version. Whether older versions are kept is up to your app or storage provider policy.
+Once you connect to the network, select your MSP, and have a bucket, the upload steps are as follows:
+
+```mermaid
+sequenceDiagram
+    participant User/App
+    participant MSP
+    participant Offchain Storage
+    participant BSP
+    participant Blockchain
+
+    User/App->>MSP: Open a storage request<br>for your bucket
+    MSP->>Offchain Storage: Accept storage request,<br>Store the bytes off-chain
+    MSP->>BSP: Coordinate replications<br>for backup copies
+    BSP->>MSP: Confirm creation<br>of replica copies
+    MSP->>Blockchain: Anchor Merkle-root<br>of bucket update
+```
+
+If replication or anchoring fails, the upload aborts.
+
+## Share Your Bucket
+
+You can share access to your bucket by generating a link that provides view-only or view-and-upload access to its contents, including all files and folders.
+
+## Retrieve the File
+
+The process to retrieve a file you successfully uploaded to DataHaven is as follows:
+
+```mermaid
+sequenceDiagram
+    participant User/App
+    participant MSP
+    participant Blockchain
+
+    User/App->>MSP: Requests the file
+    MSP->>User/App: Returns the bytes plus<br>bucket root Merkle proof
+    User/App->>Blockchain: Checks proof against<br>file's on-chain commitment
+    Blockchain->>User/App: Confirms proof check passed,<br>returned data matches original upload
+```
+
+## Verify Storage Over Time
+
+Scheduled health checks, in the form of proof challenges, ensure BSPs retain your backup copies over time as follows:
+
+```mermaid
+sequenceDiagram
+    participant Blockchain
+    participant BSP
+
+    Blockchain->>BSP: Sends proof of storage challenge as scheduled
+    BSP->>Blockchain: Provides small cryptographic proofs in response 
+```
+
+If the BSP returns the expected information, the health check passes. BSPs face slashing penalties in the event of data loss resulting in a failed proof challenge.
+
+## Change or Replace a File
+
+The flow for changing or replacing a file is as follows:
+
+```mermaid
+graph LR
+    A[Upload new content<br>to your bucket]
+    B[Change to content creates<br>updated file fingerprint]
+    C[Bucket updates on-chain<br>summary to reference<br>new version.]
+    A-->B
+    B-->C
+```
+
+It is important to note that whether older versions of files are retained is determined by your app's or storage provider's policy.
+
+## Remove Your Data
```

Comment:
I'm not sure about this section just because it's like a whole section for one sentence.